### ISSUE TITLE
Support for Reacton

### ISSUE THREAD

**Issue by rowanc1** [executablebooks/jupyterlab-myst#124]
@maartenbreddels can you help with a minimal example?

**Comment by maartenbreddels** [comment]
Python code
```python
import reacton
import reacton.ipywidgets as w

@reacton.component
def ButtonClick():
    # first render, this return 0, after that, the last argument
    # of set_clicks
    clicks, set_clicks = reacton.use_state(0)
    def my_click_handler():
        # trigger a new render with a new value for clicks
        set_clicks(clicks+1)
    button = w.Button(description=f"Clicked {clicks} times",
                      on_click=my_click_handler)
    return button
b = ButtonClick()
b
```

Markdown:
```markdown
Does it support Reacton? {eval}`b` Yup :)
```

I managed to hack it by adding the following method to out Element class:
```python
    def _repr_mimebundle_(self, **kwargs):
        box = widgets.VBox(_view_count=0)
        el = _wrap(self, jupyter_decorator_components)
        widget, rc = render(el, container=box)
        return widget._repr_mimebundle_(**kwargs)
```

But what we normally use is:
https://github.com/widgetti/reacton/blob/master/reacton/core.py#L311

This means that every time an element is displayed, it will create a new widget. This means the lifetime of the 'app' is not tied to the lifetime of the element object. I'm not sure this would match this project fully? I means that if `b` is evaluated, it would need to execute some Python code on it (e.g. display(b)), and use that output.

Does this make sense?

**Comment by maartenbreddels** [comment]
What does the current code do? Do you have a link to the code?

**Comment by rowanc1** [comment]
When the user "executes" a markdown cell, it parses the markdown and then [executes](https://github.com/executablebooks/jupyterlab-myst/blob/main/src/actions.ts#L22) any user expressions in those `{eval}` roles.

This uses the `requestExecute` method of the kernel ([here](https://github.com/executablebooks/jupyterlab-myst/blob/main/src/actions.ts#L64)), and is just a list of expressions of whatever you put between the {eval} role. This gets put in the `user_expressions` metadata on the cell ([here](https://github.com/executablebooks/jupyterlab-myst/blob/main/src/actions.ts#L90)). You can see that in the metadata of each markdown cell.

Then React gets triggered to re-render the markdown cell. Each expression is an `InlineRenderer` ([here](https://github.com/executablebooks/jupyterlab-myst/blob/main/src/inlineExpression.tsx#L155)). If the expression is not text or an error, then it uses the existing jupyter infrastructure to [create the renderer](https://github.com/executablebooks/jupyterlab-myst/blob/main/src/inlineExpression.tsx#L86) in a [react div](https://github.com/executablebooks/jupyterlab-myst/blob/main/src/inlineExpression.tsx#L152).

**Comment by agoose77** [comment]
Yes, the error here is that `ipykernel` doesn't return any mimebundles for `userExpressions`. This makes sense; `_ipython_display_` is a side-effect, which probably isn't captured by the kernel in this context.

**Comment by maartenbreddels** [comment]
With ipywidgets' Output widget we can capture any output coming from display: https://github.com/jupyter-widgets/ipywidgets/blob/93a848b5ad541db08f464a22b15d05abcfe03a63/python/jupyterlab_widgets/src/output.ts

It's a bit more involved. Also, we have a lot of issues with the output widget, but they are mostly related to multithreading, which I doubt will play a role here.

I think it would be:
 1. https://github.com/jupyter-widgets/ipywidgets/blob/93a848b5ad541db08f464a22b15d05abcfe03a63/python/jupyterlab_widgets/src/output.ts#L77
 2. execute code in Python `display(${userExpression})`
 3. handle display messages registered in step 1
 4. cleanup: https://github.com/jupyter-widgets/ipywidgets/blob/93a848b5ad541db08f464a22b15d05abcfe03a63/python/jupyterlab_widgets/src/output.ts#L72

**Comment by agoose77** [comment]
Yes, we're getting the side-effects on `iopub.display_data`, not `shell.execute_reply`:
![image](https://user-images.githubusercontent.com/1248413/227040134-4f2ffbe1-bd74-4bfc-90a2-2554a2a1a575.png)

**Comment by agoose77** [comment]
Given the above, I think we really need `reacton` to implement the mimerenderer protocol. Is there a reason not to do this?

**Comment by maartenbreddels** [comment]
Because reacton creates a widget on the fly after a call to display, it is not displayable itself (there is no widget associated to an element)

**Comment by agoose77** [comment]
Ah, I think I'm following. So, unlike `ipywidgets`, each `display()` creates a unique, independent widget that doesn't share state with previous calls to `display()`?

**Comment by maartenbreddels** [comment]
Indeed

**Comment by maartenbreddels** [comment]
```python
import reacton
import reacton.ipywidgets as w
import reacton
@reacton.component
def ButtonClick():
    # first render, this return 0, after that, the last argument
    # of set_clicks
    clicks, set_clicks = reacton.use_state(0)
    def my_click_handler():
        # trigger a new render with a new value for clicks
        set_clicks(clicks+1)
    button = w.Button(description=f"Clicked {clicks} times",
                      on_click=my_click_handler)
    return button
b = ButtonClick.widget()
b
```

This does work btw. Here we force the creation of a widget (documented only at [the solara docs](https://solara.dev/api/widget)). 
Now, we don't have a managed lifecycle, but maybe for this use case that is ok.

**Comment by maartenbreddels** [comment]
Yes, I don't think the `_repr_mimebundle_` should have 'side effects', but I'm not 100% on that. Also, I'm expecting the return value to be the same each time if I do not mutate the element (does that fall under idempotency?)

**Comment by agoose77** [comment]
I don't think there's an easy way to detect whether the object is being displayed as a cell output, or as a user expression. Also, the DAP uses user-expressions, so we need the solution to work out of the box.

I wonder if it's in-scope for the API of reacton to change, so that `ButtonClick.widget()` produces a single widget? I don't know enough about why it doesn't right now, and I'm sure it's obvious :P
